src/tests/mod.rs

```diff
@@ -0,0 +1,5 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+#![cfg(test)]
```

Comment:
Is that necessary as it is in `tests`?

Reply:
It's `src/tests`, not `./tests`. Cargo would simply think it's a normal module otherwise

Reply:
Oh I see. Any reason why it is not in `tests`?

Reply:
Yes, they're unit tests, not integration tests - they don't require any deployment or components outside of the client. Also, integration tests cannot make use of anything marked with `#[cfg(test)]`.

Reply:
I disagree about their nature, they can be seen as integration tests as they test the integration our whole stack from start to finish as opposed to individual functions 😃 
The fact that we can not use `#[cfg(test)]` should not be a factor for them to be inside `src` I think. That would support the idea I had in my suggestion.

I will not block this however because of that but might be worth thinking about in a new issue we can create about this topic!

Reply:
>I disagree about their nature, they can be seen as integration tests as they test the integration our whole stack from start to finish as opposed to individual functions

The way Rust works essentially prevents you from writing "unit" tests for this kind of functionality/library. Something that runs the whole test within the same process is the closest thing you can do to "unit testing" here. I suggest we drop the "unit" and "integration" terms as using them might give the impression that we're sticking to what's defined as such in other languages.

Also, I see no reason to take out of the source tree tests which have no need for functionality that resides outside of the crate, as long as we're clear about what these modules are. The "idiomatic" Rust way of putting "unit tests" in the same file as the source code feels way weirder and messier to me, tbh.

Reply:
(don't get me wrong, writing proper unit tests would've been my preferred way to go about this, but I'm not keen on writing the code so that we can test)